Mathlib/Combinatorics/Graph/Basic.lean

```diff
@@ -0,0 +1,323 @@
+/-
+Copyright (c) 2025 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson, Jun Kwon
+-/
+import Mathlib.Data.Set.Basic
+import Mathlib.Data.Sym.Sym2
+
+/-!
+# Multigraphs
+
+A multigraph is a set of vertices and a set of edges,
+together with incidence data that associates each edge `e`
+with an unordered pair `s(x,y)` of vertices called the *ends* of `e`.
+The pair of `e` and `s(x,y)` is called a *link*.
+The vertices `x` and `y` may be equal, in which case `e` is a *loop*.
+There may be more than one edge with the same ends.
+
+If a multigraph has no loops and has at most one edge for every given ends, it is called *simple*,
+and these objects are also formalized as `SimpleGraph`.
+
+This module defines `Graph α β` for a vertex type `α` and an edge type `β`,
+and gives basic API for incidence, adjacency and extensionality.
+The design broadly follows [Chou1994].
+
+## Main definitions
+
+For `G : Graph α β`, ...
+
+* `V(G)` denotes the vertex set of `G` as a term in `Set α`.
+* `E(G)` denotes the edge set of `G` as a term in `Set β`.
+* `G.IsLink e x y` means that the edge `e : β` has vertices `x : α` and `y : α` as its ends.
+* `G.Inc e x` means that the edge `e : β` has `x` as one of its ends.
+* `G.Adj x y` means that there is an edge `e` having `x` and `y` as its ends.
+* `G.IsLoopAt e x` means that `e` is a loop edge with both ends equal to `x`.
+* `G.IsNonloopAt e x` means that `e` is a non-loop edge with one end equal to `x`.
```

Comment:
Do you say "a nonloop" rather than "a non-loop"?
```suggestion
* `G.IsNonLoopAt e x` means that `e` is a non-loop edge with one end equal to `x`.
```